e2e: GetSafeIPAddress() replaces GetRandomIPAddress

edsantiago opened the pull request:
For tests that use '--ip XX', random IP allocation is not
working well. Switch instead to a deterministic algorithm
with CPU affinity and a fudge factor for CNI.

Fixes: #18855 

Signed-off-by: Ed Santiago <santiago@redhat.com>
```release-note
None
```

Luap99 commented on test/e2e/common_test.go:
```diff
@@ -479,19 +482,23 @@ func GetPortLock(port string) *lockfile.LockFile {
+// one test calls us more than 25 times or if some other test runs more
+// than ten networked containers at the same time as any test that
+// relies on GetSafeIPAddress(). I'm finding it hard to care.
+func GetSafeIPAddress(base *int) string {
+	*base += 10
+	return fmt.Sprintf("10.88.%d.%d", *base/256, *base%256)
```
you can just access safeIP directly, there is no need to accept this a parameter and pass it for each caller.

edsantiago replied:
Oh, excellent, thanks. So done.

Luap99 commented on test/e2e/common_test.go:
```diff
@@ -479,19 +482,23 @@ func GetPortLock(port string) *lockfile.LockFile {
+// one test calls us more than 25 times or if some other test runs more
+// than ten networked containers at the same time as any test that
+// relies on GetSafeIPAddress(). I'm finding it hard to care.
+func GetSafeIPAddress(base *int) string {
+	*base += 10
+	return fmt.Sprintf("10.88.%d.%d", *base/256, *base%256)
```
```suggestion
func GetSafeIPAddress(base *int) string {
	*base += 10
	return fmt.Sprintf("10.88.%d.%d", *base/255, *base%255)
```
Note sure if this has other side effects but you must make sure the third and fourth octet can never be 255 at the same time. 10.88.255.255 is the broadcast address and cannot be assigned to containers.

edsantiago replied:
octet 3 will be 255 iff running on ginkgo processor 127. I'm choosing to stick my head in the sand about that.

octet 4 ... its initial value is 2, and gets incremented by 10, so I'm thinking it's unlikely ever to be 255. That's specifically why I chose 256, it's a cheap way to do two octets. (Side note: I initially used uint16 but gave up after ending up with waaaaay too many cluttery typecasts). Have I missed something? I intend to keep the code as 256, but will wait to re-push; please lmk if there's something I missed.

Luap99 replied:
Yeah you make a fair point, if you think this cannot happen I am fine with it. It is to late for me to do this math in my head.

edsantiago replied:
Understood. There is no rush on this, it can wait for tomorrow or later. I've been testing heavily on my no-flake-retry PR, and (so far) no dup-IP failures, compared to at least one per run since merging #18858. My latest test run has 15 minutes to go, I'll let that finish before re-pushing this.

(Side note: "unlikely", in my comment above, was a euphemism for "impossible under the rules of the universe I live in except for cosmic-ray glitches": even number, plus `N` * another even number, can never be odd).

Luap99 commented on test/e2e/common_test.go:
```diff
@@ -106,12 +106,15 @@ var (
 		podmanTest = PodmanTestCreate(tempdir)
 		podmanTest.Setup()
+		// see GetSafeIPAddress() below
+		safeIP = (GinkgoT().ParallelProcess()+128)*256 + 2
```
maybe choose a different name, this is not a ip, it is a single integer.

edsantiago replied:
Good point, is `safeIPOctets` more sensible?

Luap99 replied:
yes

edsantiago commented on test/e2e/common_test.go:
```diff
@@ -479,19 +483,23 @@ func GetPortLock(port string) *lockfile.LockFile {
+// relies on GetSafeIPAddress(). I'm finding it hard to care.
+func GetSafeIPAddress() string {
+	safeIPOctets[1] += 10
+	return fmt.Sprintf("10.88.%d.%d", safeIPOctets[0], safeIPOctets[1])
```
`safeIPOctets...` (dot dot dot) does not work

edsantiago commented:
Converted to draft. Playing with what I think is a more readable form.

edsantiago commented:
I think I'm happy-ish with this approach. I skimmed through [f37 root logs](https://api.cirrus-ci.com/v1/artifact/task/5220390251790336/html/int-podman-fedora-37-root-host-sqlite.log.html) (CNI), searching in-page for `10.88`, and see that that CNI does indeed seem to allocate based on most-recently-used+1. So there's still a collision risk under some timings but without a way to control CNI's behavior I can find no way to prevent this.

Also, I'm much happier with the `[2]uint8` approach and am prepared to squash that for the final version.

edsantiago commented:
Sigh. Still seeing [dup-ip flakes](https://api.cirrus-ci.com/v1/artifact/task/6706881452834816/html/int-podman-fedora-37-root-host-sqlite.log.html#t--Podman-start-podman-start-container-retains-the-HOME-env-if-present--1):
```
[It] podman start container retains the HOME env if present
...
# podman [options] restart 244462f39074f7d07d229f673f4b61adbd8eedf88744bce5d9adcb037fe83697
time="2023-06-15T07:24:56-05:00" level=warning msg="Failed to load cached network config: network podman not found in CNI cache, falling back to loading network podman from disk"
Error: plugin type="bridge" failed (add): cni plugin bridge failed: failed to allocate for range 0: 10.88.129.35 has been allocated to 244462f39074f7d07d229f673f4b61adbd8eedf88744bce5d9adcb037fe83697, duplicate allocation is not allowed
```

Luap99 commented:
> Sigh. Still seeing [dup-ip flakes](https://api.cirrus-ci.com/v1/artifact/task/6706881452834816/html/int-podman-fedora-37-root-host-sqlite.log.html#t--Podman-start-podman-start-container-retains-the-HOME-env-if-present--1):
> 
> ```
> [It] podman start container retains the HOME env if present
> ...
> # podman [options] restart 244462f39074f7d07d229f673f4b61adbd8eedf88744bce5d9adcb037fe83697
> time="2023-06-15T07:24:56-05:00" level=warning msg="Failed to load cached network config: network podman not found in CNI cache, falling back to loading network podman from disk"
> Error: plugin type="bridge" failed (add): cni plugin bridge failed: failed to allocate for range 0: 10.88.129.35 has been allocated to 244462f39074f7d07d229f673f4b61adbd8eedf88744bce5d9adcb037fe83697, duplicate allocation is not allowed
> ```

This is https://github.com/containers/podman/issues/18753, I actually do not think they are the same issue.
This test never requests a static ip so it should run into this problem. The bug here is that network cleanup clearly did not work and thus CNI still sees the old ip for this container id thus rejecting to hand out a new ip for it.

Note that the errors explicitly says `duplicate allocation is not allowed` not the `requested IP address X is not available`, you are only fixing the later. The former must be a real podman bug.

edsantiago commented:
>This is https://github.com/containers/podman/issues/18753, I actually do not think they are the same issue.

I think you're right. Thank you for catching that. I've squashed, repushed, and removed the Draft marking.

Luap99 reviewed (commented):
LGTM, maybe add another comment to the function that it is recommend to create your own network if you want to test a static ip to avoid conflicts. Only tests that explicitly want to test a static ip in the default network should use this.

TomSweeneyRedHat commented:
LGTM

TomSweeneyRedHat commented:
/lgtm
Pushing this through to pickup the fix, Paul left a possible minor follow-up comment to consider @edsantiago
